google-cloud-compute/src/main/java/com/google/cloud/compute/testing/ManagedResourceCleaner.java

```
@@ -0,0 +1,210 @@
+ * limitations under the License.
+ */
+
+package com.google.cloud.compute.testing;
+
+import com.google.cloud.compute.AddressId;
+import com.google.cloud.compute.Compute;
+import com.google.cloud.compute.DiskId;
+import com.google.cloud.compute.ImageId;
+import com.google.cloud.compute.InstanceId;
+import com.google.cloud.compute.NetworkId;
+import com.google.cloud.compute.SnapshotId;
+import com.google.cloud.compute.SubnetworkId;
+import com.google.common.base.Function;
+import java.util.HashMap;
+import java.util.Map;
+import sun.reflect.generics.reflectiveObjects.NotImplementedException;
+
+public class ManagedResourceCleaner {
+  private final Compute compute;
+  private final Map<Object, ManagedResource<?>> resources;
+
+  private ManagedResourceCleaner(Compute compute) {
+    this.compute = compute;
+    this.resources = new HashMap<>();
+  }
+
+  public static ManagedResourceCleaner create(Compute compute) {
+    return new ManagedResourceCleaner(compute);
+  }
+
+  private class ManagedResource<T> {
+    final T res;
+    Function<T, Void> func;
+
+    private ManagedResource(T res) {
+      this.res = res;
+
+      @SuppressWarnings("unchecked")
+      Function<T, Void> localFunc = (Function<T, Void>) getDeleteMethod(res);
+
+      this.func = localFunc;
+    }
+
+    public void delete() {
+      func.apply(res);
+    }
+  }
+
+  public void cleanUp() {
+    for (Map.Entry<Object, ManagedResource<?>> r : resources.entrySet()) {
+      System.out.println("Clean up compute engine resource: " + r.getKey().toString());
+      r.getValue().delete();
+    }
+    resources.clear();
+  }
+
+  public ManagedResourceCleaner add(AddressId addressId) {
+    ManagedResource<AddressId> r = new ManagedResource<>(addressId);
+    resources.put(addressId, r);
+    return this;
+  }
+
+  public ManagedResourceCleaner remove(AddressId addressId) {
+    resources.remove(addressId);
+    return this;
+  }
```

You don't need an overload for `remove` - you can have just one implementation. Maybe throw an exception if the id is not found in `resources`.